src/storage/src/render/sources.rs

```diff
@@ -363,34 +364,82 @@ where
+                                                .dataflow_parameters
+                                                .storage_dataflow_max_inflight_bytes
+                                        {
+                                            let (feedback_handle, feedback_data) =
+                                                scope.feedback(Default::default());
+
+                                            (
+                                                Some(feedback_handle),
+                                                Some(persist_source::FlowControl {
+                                                    progress_stream: feedback_data,
+                                                    max_inflight_bytes:
+                                                        storage_dataflow_max_inflight_bytes,
+                                                    summary: (Default::default(), 1),
+                                                }),
+                                            )
+                                        } else {
+                                            (None, None)
+                                        };
+                                    let (stream, tok) = persist_source::persist_source_core(
+                                        scope,
+                                        id,
+                                        persist_clients,
+                                        description.ingestion_metadata,
+                                        Some(as_of),
+                                        Antichain::new(),
+                                        None,
+                                        flow_control,
+                                        // Copy the logic in DeltaJoin/Get/Join to start.
+                                        |_timer, count| count > 1_000_000,
+                                    );
+                                    (stream.as_collection(), Some(tok), feedback_handle)
+                                } else {
+                                    (Collection::new(empty(scope)), None, None)
+                                };
+                            let (upsert, health_update) = crate::render::upsert::upsert(
+                                &upsert_input.enter(scope),
+                                upsert_envelope.clone(),
+                                refine_antichain(&resume_upper),
+                                previous,
+                                previous_token,
+                                base_source_config,
+                                &storage_state.instance_context,
+                                &storage_state.dataflow_parameters,
+                            );
+
+                            // If backpressure is enabled, we probe the upsert operator's
+                            // output, which is the easiest way to extract frontier information.
+                            let upsert = match feedback_handle {
```

Comment:
Why do we need to convert a stream into a probe and then the probe back into a stream and not use the stream directly?

Reply:
this is the easiest way to get a frontier without copying the data from the output stream, i believe